### PR TITLE
Export exemplars to GCM

### PR DESCRIPTION
**Pull request by realschwa**
This PR exports a map[storage.SeriesRef]record.RefExemplar to the export function in GMP. See https://github.com/GoogleCloudPlatform/prometheus-engine/pull/308

### CODE REVIEW

**Comment by pintohutch** [comment]
@realschwa let me know when this is ready for review (i.e. tag me to review it)

File: tsdb/agent/db.go [hunk 1]
```diff
@@ -799,13 +803,14 @@ func (a *appender) AppendExemplar(ref storage.SeriesRef, l labels.Labels, e exem
 			return 0, storage.ErrExemplarLabelLength
 		}
 	}
-
-	a.pendingExamplars = append(a.pendingExamplars, record.RefExemplar{
+	refE := record.RefExemplar{
```
Comment (pintohutch) [inline tsdb/agent/db.go]: nit: `refEx` or even just `ref`?

File: tsdb/head_append.go [hunk 2]
```diff
@@ -444,14 +444,21 @@ func (a *headAppender) Commit() (err error) {
...
 
 	// No errors logging to WAL, so pass the exemplars along to the in memory storage.
+	exportExemplars := make(map[storage.SeriesRef]record.RefExemplar, len(a.exemplars))
 	for _, e := range a.exemplars {
 		s := a.head.series.getByID(chunks.HeadSeriesRef(e.ref))
 		// We don't instrument exemplar appends here, all is instrumented by storage.
 		if err := a.head.exemplars.AddExemplar(s.lset, e.exemplar); err != nil {
-			if err == storage.ErrOutOfOrderExemplar {
-				continue
+			if err != storage.ErrOutOfOrderExemplar {
```
Comment (pintohutch) [inline tsdb/head_append.go]: Why did you change this logic around?

In the interest of keeping a lean fork, we'd like to change as little as possible.

File: go.mod [hunk 3]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/Azure/azure-sdk-for-go v63.0.0+incompatible
 	github.com/Azure/go-autorest/autorest v0.11.25
 	github.com/Azure/go-autorest/autorest/adal v0.9.18
-	github.com/GoogleCloudPlatform/prometheus-engine v0.4.1
+	github.com/GoogleCloudPlatform/prometheus-engine v0.5.1-0.20221123211114-163beef5eba8
```
Comment (pintohutch) [inline go.mod]: The main reason I can think of for depending on a tagged version is to ensure the user-agent injects the proper prometheus-engine export library used. We should probably wait to tag it there before doing the final cut here.

File: tsdb/head_append.go (hunk 2 above)
Comment (realschwa) [reply tsdb/head_append.go]: I think I switched this around to account for a special case that turned out not to be so special. I can switch it back around

File: tsdb/agent/db.go (hunk 1 above)
Comment (realschwa) [reply tsdb/agent/db.go]: `refEx`

File: go.mod (hunk 3 above)
Comment (pintohutch) [reply go.mod]: @realschwa can you bump this to point to https://github.com/GoogleCloudPlatform/prometheus-engine/releases/tag/v0.6.0-rc.0?

Comment (realschwa) [reply go.mod]: @pintohutch done

**Review by pintohutch — APPROVED** [review]
